### DISCUSSION TITLE
Option to disable native reflection usage -- would PR be accepted?

### DISCUSSION THREAD

**Discussion by ddurschl-mathworks** [babel/babel#14811] [Ideas]
We're seeing a surprising performance regression of 40% in Chrome and only Chrome while using babel 7 with transform-classes. While the root cause is still unknown, I can make the code fast again by swapping:

`var hasNativeReflectConstruct = _isNativeReflectConstruct();`

to:

`var hasNativeReflectConstruct = false;`

in `_createSuper` in a single file. While we triage this, it would be great to be able to force native reflection usage off. Would a PR to add this as a second option beside "loose" for this plugin be accepted?

**Comment by JLHwung** [comment]
Here is the implementation of `isNativeReflectConstruct`

https://github.com/babel/babel/blob/0e207df2f27b7d49dff042ef4152506e4e9de06d/packages/babel-helpers/src/helpers.ts#L389-L413

If `isNativeReflectConstruct` slows down quite a bit, that it must hit the slow path in the try block, which means your environment has a `Reflect` polyfill and it does not convey that it is a polyfill (otherwise it would have set `.sham`). What's your `Reflect` polyfill?

**Reply by ddurschl-mathworks** [reply]
We're just using babel-env, no other polyfills. I don't think it's isNativeReflectConstruct that's the issue, however -- I think it's _the use of native reflect_. I can confirm this on Monday by setting it `true` and seeing if the slow returns.

**Reply by ddurschl-mathworks** [reply]
OK, I got curious and checked tonight. `true` brings back the slow. This makes sense -- if it was a naughty polyfill, I'd expect to see performance penalties in all browsers.

**Reply by ddurschl-mathworks** [reply]
Aside from the root cause here -- would a PR along these lines be accepted?

**Reply by nicolo-ribaudo** [reply]
What would the PR use instead?

**Reply by ddurschl-mathworks** [reply]
The current code is as follows:

```
      if (isNativeReflectConstruct()) {
        // NOTE: This doesn't work if this.__proto__.constructor has been modified.
        var NewTarget = GET_PROTOTYPE_OF(this).constructor;
        result = Reflect.construct(Super, arguments, NewTarget);
      } else {
        result = Super.apply(this, arguments);
      }

```
The PR would add an additional check previous to the isNativeReflectConstruct() for "did the config say no-native-reflect?", thus following the existing else block (which is correct and fast for us). It looks like this would have to be implemented twice (once here and once in the core `createSuper` helper).

**Comment by nicolo-ribaudo** [comment]
Does https://babeljs.io/docs/en/assumptions#superiscallableconstructor work for you?

**Reply by ddurschl-mathworks** [reply]
My apologies, but I don't fully understand the question. That configuration doesn't seem to be checked in either implementation of createSuper (either the inline one in the plugin or the central one in babel-helpers/src/helpers.ts .
